pkg/eventprocessor/impl/executor/impl/blockscope.go

```diff
@@ -215,6 +216,84 @@ func (bs *blockScope) StateHash(ctx context.Context, chainID tableland.ChainID)
 	return executor.NewStateHash(chainID, bs.scopeVars.BlockNumber, hash), nil
 }
 
+func (bs *blockScope) CalculateTreeLeaves(ctx context.Context, chainID tableland.ChainID) error {
+	rows, err := bs.txn.QueryContext(ctx, "select prefix, id from registry where chain_id = ?1 ORDER BY rowid", chainID)
+	if err != nil {
+		return fmt.Errorf("fetching tables from registry: %s", err)
+	}
+	defer func() {
+		_ = rows.Close()
+	}()
+
+	for rows.Next() {
+		var tablePrefix string
+		var tableID int
+		if err := rows.Scan(&tablePrefix, &tableID); err != nil {
+			return fmt.Errorf("scanning table name: %s", err)
+		}
+
+		if err := bs.calculateTreeLeavesForTable(ctx, chainID, tablePrefix, tableID); err != nil {
+			return fmt.Errorf("calculate leaves for table: %s", err)
+		}
+	}
+
+	return nil
+}
+
+func (bs *blockScope) calculateTreeLeavesForTable(
+	ctx context.Context,
+	chainID tableland.ChainID,
+	tablePrefix string,
+	tableID int,
+) error {
+	tableName := fmt.Sprintf("%s_%d_%d", tablePrefix, chainID, tableID)
+	tableRows, err := bs.txn.QueryContext(ctx, fmt.Sprintf("SELECT * FROM %s", tableName))
+	if err != nil {
+		return fmt.Errorf("fetching rows from %s: %s", tableName, err)
+	}
+	defer func() {
+		_ = tableRows.Close()
+	}()
+
+	cols, err := tableRows.Columns()
+	if err != nil {
+		return fmt.Errorf("getting the columns of row: %s", err)
+	}
+
+	rawBuffer := make([]sql.RawBytes, len(cols))
+	scanCallArgs := make([]interface{}, len(rawBuffer))
+	for i := range rawBuffer {
+		scanCallArgs[i] = &rawBuffer[i]
+	}
+
+	leaves := []byte{}
+	for tableRows.Next() {
+		if err := tableRows.Scan(scanCallArgs...); err != nil {
+			return fmt.Errorf("table row scan: %s", err)
+		}
+
+		rowHash := sha3.New256()
+		for _, col := range rawBuffer {
+			rowHash.Write(col)
+		}
+
+		leaves = append(leaves, rowHash.Sum(nil)...)
+	}
+
+	if _, err := bs.txn.ExecContext(ctx,
+		"INSERT INTO system_tree_leaves (prefix, chain_id, table_id, block_number, leaves) VALUES (?1, ?2, ?3, ?4, ?5)",
```

Comment:
Here's the "snapshotting". We're copying a table into an entire row as a stream of bytes. The information will be picked up by a background process. 

Not really sure how that will fully work, but one approach use this table a queue. So the background process pick rows form here, process, then deletes the row. With that approach the process would have to keep a copy into another table or database. Why? Because it needs to generate proofs from this data at any given time. 

Another approach is to put a flag in this table to indicate that it was processed. We could prune older rows. 

In my mind what sounds like a good idea is the first delete approach combined with the serialization of the merkle tree. Here's how it would work: the background process pick a row to process, builds a tree, serialize the tree into another table (could even be another database, maybe `trees.db`. Another db would make it harder to make it transactional), then published the root to the registry. So we would have the entire tree ready in the db, when the request for a proof comes.